Add MakeSignature and VerifySignature calls

bodgit opened the pull request:
I'm in the process of making some existing GSSAPI code use your SSPI library on Windows. I've used your Negotiate library to establish a context successfully however I now need to use `MakeSignature` & `VerifySignature` and noticed they weren't present.

This PR is the result of me updating `syscall.go` and running `go generate`. I'm not quite sure how/where I'm going to use these either adding them somewhere to your Negotiate library or just calling them from my side with the established context but this should be a start.

alexbrainman commented on zsyscall_windows.go:
```diff
@@ -109,3 +111,15 @@ func ApplyControlToken(context *CtxtHandle, input *SecBufferDesc) (ret syscall.E
+	return
+}
+
+func ValidateSignature(context *CtxtHandle, message *SecBufferDesc, messageseqno uint32, qop *uint32) (ret syscall.Errno) {
```
This function is called ValidateSignature, but syscall.go says:

//sys	VerifySignature ...

How is that possible, if you run `go generate` command?

bodgit replied:
I don't normally develop on Windows and although I can cross compile for Windows from a Linux or OS X host I couldn't work out how to run `go generate` without being on an actual Windows host so I'm afraid I just ran the `go generate` and copied the diff of `zsyscall_windows.go` across. The GSSAPI equivalent function is `gss_verify_mic()` so I've gotten my verify and validate mixed up, apologies.

When I ran this with go 1.9.x it made more changes to the file other than just add these new functions but I figured that would probably break compatibility with older golang's so I haven't included them.

alexbrainman replied:
> I don't normally develop on Windows and although I can cross compile for Windows from a Linux or OS X host I couldn't work out how to run go generate without being on an actual Windows host

`go generate` only worked on Windows. But now I moved it to a file that is visible on any OS, so you should be able to use it too. I also added your functions here https://github.com/alexbrainman/sspi/commit/bc15a409b6322b891e172a20c0bb8f0924b0cda2

bodgit commented:
I've started trying to use these new calls and I've found I need to call `sspi.QueryContextAttributes()` with `_SECPKG_ATTR_SIZES` to work out how large a buffer I need to be able to store the signature I'm about make.

The `schannel` package has all of the various constants defined as well as functions in `schannel/attribute.go` to query some of the other `_SECPKG_ATTR_*` values but AFAIK these can't be called with my `negotiate.ClientContext`. Do some of these functions perhaps need shuffling about? I was thinking that it might make sense to move the main function to the `sspi` package and then add thin wrappers to each of the relevant SSP.

For example, define in `sspi`:
```golang
const (
        _SECPKG_ATTR_SIZES = 0
)

type _SecPkgContext_Sizes struct {
       MaxToken        uint32
       MaxSignature    uint32
       BlockSize       uint32
       SecurityTrailer uint32
}

func (c *Context) Sizes() (uint32, uint32, uint32, uint32, error) {
       var s _SecPkgContext_Sizes
       ret := QueryContextAttributes(c.Handle, _SECPKG_ATTR_SIZES, (*byte)(unsafe.Pointer(&s)))
       if ret != SEC_E_OK {
               return 0, 0, 0, 0, ret
       }
       return s.MaxToken, s.MaxSignature, s.BlockSize, s.SecurityTrailer, nil
}
```
And then in `sspi.negotiate` just have:
```golang
func (c *ClientContext) Sizes() (uint32, uint32, uint32, uint32, error) {
       return sspi.Sizes(c.sctxt)
}

func (c *ServerContext) Sizes() (uint32, uint32, uint32, uint32, error) {
       return sspi.Sizes(c.sctxt)
}
```
etc.

alexbrainman commented:
> For example, define in sspi:
> 
> const (
>         _SECPKG_ATTR_SIZES = 0
> )

I don't think SECPKG_ATTR_SIZES is part of SSPI. I think it (and others) are part of schannel Microsoft package. Just copy bits that you need into your code. Will that work for you?

Alex

bodgit commented:
> I don't think SECPKG_ATTR_SIZES is part of SSPI. I think it (and others) are part of schannel Microsoft package.

If you compare the documentation for `QueryContextAttributes` between say the [negotiate version](https://msdn.microsoft.com/en-us/library/windows/desktop/aa379332(v=vs.85).aspx) and the [schannel version](https://msdn.microsoft.com/en-us/library/windows/desktop/aa379340(v=vs.85).aspx) you can see the list of attributes implemented by both. The schannel version for example implements the SECPKG_ATTR_STREAM_SIZES attribute but the negotiate version doesn't, however both implement the SECPKG_ATTR_SIZES attribute. So I think it would make sense to move these to the base SSPI package and then add the wrappers to each security package where that particular attribute is implemented, thankfully the docs should make this straightforward.

> Just copy bits that you need into your code. Will that work for you?

Unfortunately golang prohibits me from accessing the `negotiate.ClientContext.sctxt` member so I can't do much from outside of your SSPI package.

Happy to extend this PR if you want.

bodgit commented:
This last commit adds everything that I needed to get things working. I added an exported `NewSecBufferDesc` function based on the unexported version in the `schannel` package as that seemed useful to have and added `MakeSignature` and `VerifySignature` functions to the `negotiate` package, (according to the docs this isn't supported by the Schannel SSP so I've not added them there).

The library that I've written is here: https://github.com/bodgit/tsig which allows me to do GSS-TSIG signed dynamic DNS updates (primarily to Windows servers) from either Windows clients using SSPI, or from non-Windows clients using GSSAPI). The negotiate handshake itself is performed over DNS using TKEY records.

alexbrainman commented on sspi.go:
```diff
@@ -152,3 +152,20 @@ func (c *Context) RevertToSelf() error {
 	return nil
 }
+
+func (c *Context) Sizes() (uint32, uint32, uint32, uint32, error) {
```
Needs documentation.

alexbrainman commented on sspi.go:
```diff
@@ -152,3 +152,20 @@ func (c *Context) RevertToSelf() error {
+	return s.MaxToken, s.MaxSignature, s.BlockSize, s.SecurityTrailer, nil
+}
+
+func NewSecBufferDesc(b []SecBuffer) *SecBufferDesc {
```
Needs documentation.

alexbrainman commented on schannel/attribute.go:
```diff
@@ -72,3 +72,7 @@ func (c *Client) KeyInfo() (sessionKeySize uint32, sigAlg uint32, sigAlgName str
 	return ki.KeySize, ki.SignatureAlgorithm, saname, ki.EncryptAlgorithm, eaname, nil
 }
+
+func (c *Client) Sizes() (uint32, uint32, uint32, uint32, error) {
```
Needs documentation.

alexbrainman commented on negotiate/negotiate.go:
```diff
@@ -263,3 +312,15 @@ func (c *ServerContext) ImpersonateUser() error {
 	return c.sctxt.RevertToSelf()
 }
+
+func (c *ServerContext) Sizes() (uint32, uint32, uint32, uint32, error) {
```
Needs documentation.

alexbrainman commented on negotiate/negotiate.go:
```diff
@@ -263,3 +312,15 @@ func (c *ServerContext) ImpersonateUser() error {
+	return c.sctxt.Sizes()
+}
+
+func (c *ServerContext) MakeSignature(msg []byte, qos, seqno uint32) ([]byte, error) {
```
Needs documentation.

alexbrainman commented on negotiate/negotiate.go:
```diff
@@ -263,3 +312,15 @@ func (c *ServerContext) ImpersonateUser() error {
+	return makeSignature(c.sctxt, msg, qos, seqno)
+}
+
+func (c *ServerContext) VerifySignature(msg, token []byte, seqno uint32) (uint32, error) {
```
Needs documentation.

alexbrainman commented on negotiate/negotiate.go:
```diff
@@ -193,6 +230,18 @@ func (c *ClientContext) Update(token []byte) (authCompleted bool, outputToken []
 	return authDone, otoken, nil
 }
 
+func (c *ClientContext) Sizes() (uint32, uint32, uint32, uint32, error) {
```
Needs documentation.

alexbrainman commented on negotiate/negotiate.go:
```diff
@@ -193,6 +230,18 @@ func (c *ClientContext) Update(token []byte) (authCompleted bool, outputToken []
+	return c.sctxt.Sizes()
+}
+
+func (c *ClientContext) MakeSignature(msg []byte, qos, seqno uint32) ([]byte, error) {
```
Needs documentation.

alexbrainman commented on negotiate/negotiate.go:
```diff
@@ -193,6 +230,18 @@ func (c *ClientContext) Update(token []byte) (authCompleted bool, outputToken []
+	return makeSignature(c.sctxt, msg, qos, seqno)
+}
+
+func (c *ClientContext) VerifySignature(msg, token []byte, seqno uint32) (uint32, error) {
```
Needs documentation.

bodgit commented:
Fixed. Incidentally, I noticed the `testContextExpiry` test always fails for me; the context expiry always seems to be around 5 minutes rather than longer than an hour.

alexbrainman commented:
> Incidentally, I noticed the testContextExpiry test always fails for me; the context expiry always seems to be around 5 minutes rather than longer than an hour.

I don't have much free time to spend investigating bugs in this packages, but changes are welcome.

Alex
